kubevirtexternalinfra/externalinfra.go

```diff
@@ -0,0 +1,98 @@
+package kubevirtexternalinfra
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1beta1"
+)
+
+type KubevirtInfraClientMap interface {
+	DiscoverKubevirtClusterClient(context.Context, client.Client, *hyperv1.HostedCluster, string) (*KubevirtInfraClient, error)
+	Delete(string)
+}
+
+func NewKubevirtInfraClientMap() KubevirtInfraClientMap {
+	return &kubevirtInfraClientMapImp{
+		theMap: sync.Map{},
+	}
+}
+
+type kubevirtInfraClientMapImp struct {
+	theMap sync.Map
+}
+
+type KubevirtInfraClient struct {
+	client.Client
+	Namespace string
+}
+
+func (k *kubevirtInfraClientMapImp) DiscoverKubevirtClusterClient(ctx context.Context, cl client.Client, hc *hyperv1.HostedCluster, localInfraNamespace string) (*KubevirtInfraClient, error) {
+	if hc.Spec.Platform.Kubevirt.Credentials == nil {
+		return &KubevirtInfraClient{
+			Client:    cl,
+			Namespace: localInfraNamespace,
+		}, nil
+	}
+	loaded, ok := k.theMap.Load(hc.Spec.InfraID)
+	if ok {
+		return loaded.(*KubevirtInfraClient), nil
+	}
+	targetClient, err := generateKubevirtInfraClusterClient(ctx, cl, *hc)
+	if err != nil {
+		return nil, err
+	}
+
+	cluster := &KubevirtInfraClient{
+		Client:    targetClient,
+		Namespace: hc.Spec.Platform.Kubevirt.Credentials.InfraNamespace,
+	}
+
+	k.theMap.LoadOrStore(hc.Spec.InfraID, cluster)
```

Comment:
since this is abstracted for both HC and NP controllers now, you could have this `DiscoverKubevirtClusterClient` function take in a `key`. That way we don't have to depend on the HC infraID always being the key (which might cause issues when hc is somehow deleted before np)